classes/actions.php

```diff
@@ -208,10 +208,22 @@ public static function duplicate_to_course(array $modules, int $targetcourseid,
         $sourcemodinfo = get_fast_modinfo($sourcecourseid);
         $targetmodinfo = get_fast_modinfo($targetcourseid);
 
-        // If a the target section number has been specified and there is no such target section, we do nothing.
-        // This should have been prevented by the UI anyway, but we want to be extra safe here.
+        // If a new section has been specified, we create one.
         if ($sectionnum != -1 && $sectionnum > count($targetmodinfo->get_section_info_all()) - 1) {
-            return;
+            $course = $DB->get_record('course', array('id' => $targetcourseid), '*', MUST_EXIST);
+            $courseformat = course_get_format($course);
+            $courseformatoptions = $courseformat->get_format_options();
+
+            // Ensure a legitimate sectionnum is given.
+            if (isset($courseformatoptions['numsections'])) { // Course format setting.
+                $sectionnum = $courseformatoptions['numsections'] + 1;
+                // Update format setting.
+                update_course((object)array('id' => $course->id,
+                'numsections' => $sectionnum));
+            } else if (course_get_format($course)->uses_sections()) { // No format setting.
+                $sectionnum = $courseformat->get_last_section_number() + 1;
+            }
```

Comment:
I do not really understand why we need to update course settings here. Isn't this what `course_create_section` does?

If we have a sectionnum greater than current count of sections (which is the case in this if block) we just can create a new section with `$newsection = course_create_section($targetcourseid)->sectionnum;`, can't we?
Alternatively, we just can manually set `$sectionnum = $targetmodinfo->get_section_info_all() + 1`?

I'm not super deep into this stuff, so you might have your reasons here :-)

Reply:
I looked at how it was done in course/changenumsections.php.  Some course formats have a setting for the amount of sections.  If that setting isn't also changed, the items are duplicated into a section, but the section is considered "orphaned".  This is also the case for when we need to create multiple sections to match the original course.  The setting needs updated to prevent orphaned sections.

Reply:
OK, well done. This definitely feels that there should be a core API function dealing with these kinds of things, but does not look like one exists...